src/coreclr/src/jit/earlyprop.cpp

```diff
@@ -477,162 +485,170 @@ GenTree* Compiler::optPropGetValueRec(unsigned lclNum, unsigned ssaNum, optPropK
-
     if ((compCurBB->bbFlags & BBF_HAS_NULLCHECK) == 0)
     {
         return;
     }
 
-    assert(tree->OperIsIndir());
+    GenTree*   nullCheckTree   = optFindNullCheckToFold(tree, nullCheckMap);
+    GenTree*   nullCheckParent = nullptr;
+    Statement* nullCheckStmt   = nullptr;
+    if ((nullCheckTree != nullptr) && optIsNullCheckFoldingLegal(tree, nullCheckTree, &nullCheckParent, &nullCheckStmt))
+    {
+#ifdef DEBUG
+        if (verbose)
+        {
+            printf("optEarlyProp Marking a null check for removal\n");
+            gtDispTree(nullCheckTree);
+            printf("\n");
+        }
+#endif
+        // Remove the null check
+        nullCheckTree->gtFlags &= ~(GTF_EXCEPT | GTF_DONT_CSE);
 
-    GenTree* const addr = tree->AsIndir()->Addr();
-    if (addr->OperGet() == GT_LCL_VAR)
+        // Set this flag to prevent reordering
+        nullCheckTree->gtFlags |= GTF_ORDER_SIDEEFF;
+        nullCheckTree->gtFlags |= GTF_IND_NONFAULTING;
+
+        nullCheckParent->gtFlags &= ~GTF_DONT_CSE;
+
+        nullCheckMap->Remove(nullCheckTree->gtGetOp1()->AsLclVarCommon()->GetLclNum());
+
+        // Re-morph the statement.
+        Statement* curStmt = compCurStmt;
+        fgMorphBlockStmt(compCurBB, nullCheckStmt DEBUGARG("optFoldNullCheck"));
+        compCurStmt = curStmt;
+    }
+
+    if ((tree->OperGet() == GT_NULLCHECK) && (tree->gtGetOp1()->OperGet() == GT_LCL_VAR))
+    {
+        nullCheckMap->Set(tree->gtGetOp1()->AsLclVarCommon()->GetLclNum(), tree,
+                          LocalNumberToNullCheckTreeMap::SetKind::Overwrite);
+    }
+}
+
+//----------------------------------------------------------------
+// optFindNullCheckToFold: Try to find a GT_NULLCHECK node that can be folded into the indirection node.
+//
+// Arguments:
+//    tree           - The input indirection tree.
+//    nullCheckMap   - Map of the local numbers to the latest NULLCHECKs on those locals in the current basic block
+//
+// Notes:
+//    Check for cases where
+//    1. One of the following trees
+//
+//       nullcheck(x)
+//       or
+//       x = comma(nullcheck(y), add(y, const1))
+//
+//       is post-dominated in the same basic block by one of the following trees
+//
+//       indir(x)
+//       or
+//       indir(add(x, const2))
+//
+//       (indir is any node for which OperIsIndirOrArrLength() is true.)
+//
+//     2.  const1 + const2 if sufficiently small.
+
+GenTree* Compiler::optFindNullCheckToFold(GenTree* tree, LocalNumberToNullCheckTreeMap* nullCheckMap)
+{
+    assert(tree->OperIsIndirOrArrLength());
+
+    GenTree* addr = (tree->OperGet() == GT_ARR_LENGTH) ? tree->AsArrLen()->ArrRef() : tree->AsIndir()->Addr();
+
+    ssize_t offsetValue = 0;
+
+    if ((addr->OperGet() == GT_ADD) && addr->gtGetOp2()->IsCnsIntOrI())
     {
-        // Check if we have the pattern above and find the nullcheck node if we do.
+        offsetValue += addr->gtGetOp2()->AsIntConCommon()->IconValue();
+        addr = addr->gtGetOp1();
```

Comment:
Maybe early exit here for large offsets?

Reply:
Will change tomorrow.

Reply:
I decided not to do that: large offsets are rare and it's better to check in the end when we have the full offset.